Initial implementation

sbor23 opened the pull request:
Tests only ensure that a full document *can* be parsed.
The actual client part is missing still, but parsing to models should
work.

Commit:
Initial implementation

Tests only ensure that a full document *can* be parsed.
The actual client part is missing still, but parsing to models should
work.

sbor23 commented:
Will setup Github Actions once this goes through. :)

czosel commented on Pipfile:
```diff
@@ -0,0 +1,15 @@
+requests = "*"
+
+[requires]
+python_version = "3.6"
```
nice, pipenv! :tada:

czosel commented on Untitled.ipynb:
```diff
@@ -0,0 +1,207 @@
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
do we need this file?

sbor23 replied:
ofc not

czosel commented:
Probably @anehx should also take a quick look.

czosel commented on caluma_client/tests/test_parser.py:
```diff
@@ -0,0 +1,13 @@
+    data = json.load(response_file.open())
+
+    raw = parse_document(data, nesting="data.allDocuments.edges.0.node")
+    Document(raw)
```
Would it be possible to serialize the parsing result somehow? (and use snapshot testing on it?)

sbor23 replied:
So you suggest we do a kind of round-trip testing (JSON -> models -> JSON) and verify using snapshots? It should be possible as I plan to implement visitors anyway, and the generic visitor could just print `self.raw` for each of the models.

czosel replied:
Sounds good! I'd do it only if it's simple to implement and if it helps while development.

open-dynaMIX commented on caluma_client/models.py:
```diff
@@ -0,0 +1,179 @@
+    "DynamicMultipleChoiceQuestion": "ListAnswer",
+    "DynamicChoiceQuestion": "StringAnswer",
+    "TableQuestion": "TableAnswer",
+    "FormQuestion": "FormAnswer",
```
There is no `FormAnswer`, this should be `None` as well.

sbor23 replied:
Hmm ok this is adapted from https://github.com/projectcaluma/ember-caluma/blob/master/addon/lib/field.js#L41
Maybe I should suggest this in ember-caluma as well.

open-dynaMIX replied:
That would be great!

There was a `FormAnswer` once, but it broke everything (nested Answers). So we removed it.

sbor23 replied:
Done

ganwell commented on caluma_client/models.py:
```diff
@@ -0,0 +1,179 @@
+
+    @property
+    def uuid(self):
+        return decode_id(self.raw["id"]) if "id" in self.raw else None
```
I like that approach. Maybe something like this would make creating @properties faster:

```python
def __init__(self....):
    ...
    self.uuid = create_getter(type='uuid', dom_path='id', default=None)
    ...

create_getter returns a closure that returns the right data from raw.

sbor23 replied:
The `uuid` and other fields with `@property` are more or less a left-over from porting the `ember-caluma/addon/lib` code. There, a lot of fields need to be dynamically reloadable when there are updates. But since this library doesn't do updates, we could purge the `@property` usages altogether.

ganwell replied:
I my idea was more about generating path accessors and conversions, so the definition of properties / fields or - what ever you want to call it - is stictly declarative and the logic is completely hidden in create_getter. Then it feels like a schema definition. It all depends how many different types of fields and how complicated the dom-path can become. It might be total overkill or not... dunno.

sbor23 replied:
That was addressed in the recent commit

ganwell commented on caluma_client/models.py:
```diff
@@ -0,0 +1,179 @@
+        self.fieldset = fieldset
+        self.raw = raw
+
+        self.question = Question(self.raw["question"])
```
Would it make sense to make this lazy? see also create_getter() below.

sbor23 replied:
That would be very cool, thanks for the recommendation. :)

Something like this maybe?
```python
    ...
    self.question = self._create_question()

def _create_question(self):
    return lambda: Question(self.raw["question"])
```

ganwell replied:
Yes. Exactly

If you'd implement something like the proposed create_getter above, you can hide the lazyness in there.

sbor23 replied:
Solved this using `make_property`

open-dynaMIX reviewed (changes requested):
Awesome work :tada:

Just a nitpick.

Commit:
refactor: address code review feedback

Make most fields lazy properties and get rid of the using
`self._create_xxx` antipattern in constructors.

open-dynaMIX reviewed (approved):
LGTM :tada:
